File: 876ex.js
'use strict';

/**
 * HOW TO USE:
 * 
 * const createApiClient = require('876ex.js');
 * 
 * var client = createApiClient('your-api-key', 'your-api-secret', {
 *     host: 'api.876ex.com',
 *     https: true,
 *     timeout: 5,
 *     debug: true
 * });
 * 
 * client.get('/v1/users/wss/token', { k1: 'v1', k2: 'v2' }, function (err, resp) {
 *     if (err) {
 *         console.error('ERROR: ' + err);
 *     } else {
 *         console.log(resp)
 *     }
 * });
 * 
 * client.post('/v1/wallet/withdraw/addresses', { addressCurrency: 'BTC', address: '1BTCxxx', description: 'test' }, function (err, resp) {
 *     if (err) {
 *         console.error('ERROR: ' + err);
 *     } else {
 *         console.log(resp)
 *     }
 * });
 */
const
    http = require('http'),
    https = require('https'),
    crypto = require('crypto');

function ApiError(error, data, message) {
    this.error = error;
    this.data = data;
    this.message = message;
}

ApiError.prototype.name = 'ApiError';
ApiError.prototype.constructor = ApiError;
ApiError.prototype.toString = function () {
    return 'ApiError(' + this.error + ', data=' + this.data + ', message=' + this.message + ')';
};

/**
 * create API client.
 * 
 * @param {string} apiKey the api key.
 * @param {string} apiSecret the api secret.
 * @param {object} options options like { host: 'api.876ex.com', https: true, timeout: 5, debug: true }
 */
function createApiClient(apiKey, apiSecret, options) {
    var
        host = (options.host || 'api.876ex.com').toLowerCase(),
        protocol = options.https === false ? 'http' : 'https',
        timeout = options.timeout || 10,
        debug = options.debug || false;
    return new ApiClient(apiKey, apiSecret, host, protocol, timeout, debug);
}

function ApiClient(apiKey, apiSecret, host, protocol, timeout, debug) {
    var pos = host.indexOf(':');
    this.apiKey = apiKey;
    this.apiSecret = apiSecret;
    this.host = host;
    this.hostname = pos > 0 ? host.substring(0, pos) : host;
    this.port = pos > 0 ? host.substring(pos + 1) : (protocol === 'https' ? 443 : 80);
    this.protocol = protocol;
    this.timeout = timeout;
    this.debug = debug;
}

ApiClient.prototype.get = function (path, params, callback) {
    if (arguments.length === 2) {
        callback = params;
        params = null;
    }
    var
        k,
        ps = [];
    if (params) {
        for (k in params) {
            ps.push(k + '=' + params[k]);
        }
        ps.sort();
    }
    this._http('GET', path, ps.join('&'), '', callback);
}

ApiClient.prototype.post = function (path, data, callback) {
    if (arguments.length === 2) {
        callback = data;
        data = {};
    }
    this._http('POST', path, '', JSON.stringify(data || {}), callback);
}

ApiClient.prototype._http = function (method, path, params, data, callback) {
    var
        payload, str, sign, req, k, url,
        ts = Date.now(),
        lines = [method, this.hostname, path, params],
        headers = {
            'API-KEY': this.apiKey,
            'API-SIGNATURE-METHOD': 'HmacSHA256',
            'API-SIGNATURE-VERSION': '1',
            'API-TIMESTAMP': '' + ts,
            'API-UNIQUE-ID': 'uk' + ts
        },
        headerLines = [];
    for (k in headers) {
        headerLines.push(k + ': ' + headers[k]);
    }
    headerLines.sort();
    payload = lines.concat(headerLines);
    payload.push(data ? data : '')
    console.log(JSON.stringify(payload))
    str = payload.join('\n');
    sign = crypto.createHmac('sha256', this.apiSecret).update(str).digest('hex');
    if (this.debug) {
        console.log('payload:\n----\n' + str + '----\nsignature: ' + sign);
    }
    // build request:
    headers['API-SIGNATURE'] = sign;
    if (method === 'GET' && params) {
        path = path + '?' + params;
    }
    if (method === 'POST') {
        headers['Content-Type'] = 'application/json';
    }
    url = this.protocol + '://' + this.host + path;
    req = (this.protocol === 'https' ? https : http).request({
        host: this.hostname,
        port: this.port,
        method: method,
        path: path,
        headers: headers,
    }, function (res) {
        res.on('data', function (data) {
            var result = JSON.parse(data.toString('utf8'));
            if (result.error) {
                callback(new ApiError(result.error, result.data, result.message));
            } else {
                callback(null, result);
            }
        });
    });
    if (method === 'POST') {
        req.write(data);
    }
    req.end();
}

if (process && process.version) {
    // nodejs
    module.exports = createApiClient;
}

File: 876ex.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import json, time, hmac, uuid, urllib, base64, hashlib

from urllib import request, parse, error

DEFAULT_HOST = 'api.876ex.com'

class ApiClient(object):

    def __init__(self, api_key, api_secret, host=None, https=True, timeout=10, debug=False):
        self._api_key = api_key
        self._api_secret = api_secret.encode('utf-8')
        self._host = (host or DEFAULT_HOST).lower()
        self._protocol = 'https' if https else 'http'
        self._timeout = timeout
        self._debug = debug

    def _hostname(self):
        n = self._host.find(':')
        if n > 0:
            return self._host[:n]
        return self._host

    def get(self, path, **params):
        return self._http('GET', path, params, None)

    def post(self, path, obj=None):
        data = json.dumps(obj) if obj is not None else None
        return self._http('POST', path, {}, data)

    def _http(self, method, path, params, data):
        # build payload:
        param_list = ['%s=%s' % (k, v) for k, v in params.items()]
        param_list.sort()
        payload = [method, self._hostname(), path, '&'.join(param_list)]
        headers = {
            'API-Key': self._api_key,
            'API-Signature-Method': 'HmacSHA256',
            'API-Signature-Version': '1',
            'API-Timestamp': str(int(time.time() * 1000))
        }
        if method == 'POST':
            headers['API-Unique-ID'] = uuid.uuid4().hex
        headers_list = ['%s: %s' % (k.upper(), v) for k, v in headers.items()]
        headers_list.sort()
        payload.extend(headers_list)
        payload.append(data if data else '')
        payload_str = '\n'.join(payload)
        # signature:
        sign = hmac.new(self._api_secret, payload_str.encode('utf-8'), hashlib.sha256).hexdigest()
        if self._debug:
            self.debug('payload:\n----\n' + payload_str + '----\nsignature: ' + sign)
        headers['API-Signature'] = sign
        # build request:
        if data:
            data = data.encode('utf-8')
        else:
            data = None
        url = '%s://%s%s?%s' % (self._protocol, self._host, path, parse.urlencode(params))
        req = request.Request(url, data=data, method=method)
        for k, v in headers.items():
            req.add_header(k, v)
        if data:
            req.add_header('Content-Type', 'application/json')
        if self._debug:
            self.debug('%s: %s' % (method, url))
            curl = 'curl -H \'Content-Type: application/json\''
            for k, v in headers.items():
                curl = curl + ' -H \'%s: %s\'' % (k, v)
            if method == 'POST':
                curl = curl + ' -d \'%s\'' % data
            curl = curl + ' ' + url
            self.debug(curl)
        try:
            with request.urlopen(req, timeout=self._timeout) as f:
                s = f.read()
                r = json.loads(s.decode('utf-8'), object_hook=lambda d: Dict(**d))
                if self._debug:
                    self.debug('Response:\n' + json.dumps(r))
                return r
        except error.HTTPError as err:
            s = err.read()
            if self._debug:
                self.debug(s)
            return json.loads(s.decode('utf-8'), object_hook=lambda d: Dict(**d))

    def debug(self, msg):
        print(msg)

class ApiError(Exception):
    pass

class Dict(dict):

    def __init__(self, **kw):
        super().__init__(**kw)

    def __getattr__(self, key):
        try:
            return self[key]
        except KeyError:
            raise AttributeError(r"'Dict' object has no attribute '%s'" % key)

    def __setattr__(self, key, value):
        self[key] = value
